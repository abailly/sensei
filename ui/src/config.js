export const config = {
  user: USER
}
