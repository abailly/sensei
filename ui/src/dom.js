/**
 * @fileOverview
 * @name dom.js
 * @author arnaud@pankzsoft.com
 * @license MIT
 */



/**
 * Basic implementation of JSX element transformation function.
 * Stolen from <a href='https://blog.r0b.io/post/using-jsx-without-react/'>this blog post</a>.
 *
 * @param {String} tagName the name of the tag to create
 * @param {Object} attrs an object defining all the attributes of the new node
 * @param {any} children a possibly empty sequence of children to add to this node.
 * @returns {Object} a newly initialized DOM element with given attributes and children
 */
export function dom(tagName, attrs = {}, ...children) {
  const elem = Object.assign(document.createElement(tagName), attrs);
  for (const child of children) {
    if (Array.isArray(child)) elem.append(...child);
    else elem.append(child);
  }
  return elem;
}

export function clear(parentId) {
  const timelines = document.getElementById(parentId);
  clearElement(timelines);
}

export function clearElement(elem) {
  while (elem.firstChild) {
    elem.removeChild(elem.firstChild);
  }
}
