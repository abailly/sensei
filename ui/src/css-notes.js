import {dom} from './dom.js';
import {ChronoUnit, LocalDateTime, LocalTime} from "@js-joda/core";
import {config} from "./config";
import showdown from "showdown";
import {formatISODateTime} from "./date";

const THIRTY_MINUTES = 30;
const HALF_HOUR_WIDTH = 100;

export function drawCssNotes(container, day, notesData) {
    let previousTime = LocalTime.parse(config.userProfile.userStartOfDay);
    const mainNode = document.getElementById("main");
    const mainParentNode = mainNode.parentNode;

    const timelineContainer = container.firstChild;
    const notesHeader = <div id={'title-notes-' + day} class="timeline-header">Notes</div>;
    const notesDiv = <div id={'notes-' + day} class="timeline-chart"/>;
    const noteList = <ul/>;


    clearNotes();
    drawNotes();
    notesDiv.appendChild(noteList);
    timelineContainer.insertBefore(notesHeader, timelineContainer.children[1]);
    timelineContainer.insertBefore(notesDiv, timelineContainer.children[2]);

    function clearNotes() {
        if(document.getElementById('title-notes-' + day)) {
            document.getElementById('title-notes-' + day).remove();
        }
        notesData.map(note => document.getElementById('note-' + note.noteStart))
            .filter(node => node !== null)
            .forEach(node => node.remove());
        notesData.map(note => document.getElementById('notes-' + day))
            .filter(node => node !== null)
            .forEach(node => node.remove());
    }

    function drawNotes() {
        notesData.forEach(note => {
            noteList.appendChild(drawNote(note));
        });
    }

    function drawNote(note) {
        const noteDisplay = <li style={'width:16px; margin-left:' + noteLeftMargin(note) + 'px;'}/>;
        const dialog = <div id={'note-' + note.noteStart} class="c-dialog"/>;
        const dialogBox = <div role="document" class="c-dialog__box"/>;
        const closeButton = <div>X</div>;
        const dialogContent = <div id={'note-desc-' + note.noteStart}/>;
        const button = <div type="button" id={'note-button-' + note.noteStart}><i /></div>;

        noteDisplay.appendChild(button);
        dialog.appendChild(dialogBox);
        dialogBox.appendChild(closeButton);
        dialogBox.appendChild(dialogContent);
        dialogContent.insertAdjacentHTML("beforeend", formatNote(note));
        mainParentNode.insertBefore(dialog, mainNode.nextSibling);

        const setAttributesTo = (node, values) => {
            values.forEach(value => node.setAttribute(value.type, value.value));
        }

        setAttributesTo(dialog, [
            {type: "role", value: dialog},
            {type: "aria-describedby", value: 'note-desc-' + note.noteStart},
            {type: "aria-modal", value: 'true'},
            {type: "aria-hidden", value: 'true'},
            {type: "tabindex", value: '-1'}
        ]);

        setAttributesTo(closeButton, [
            {type: "type", value: "button"},
            {type: "aria-label", value: "Close"},
            {type: "aria-dismiss", value: "dialog"},
        ]);

        setAttributesTo(button, [
            {type: "type", value: "button"},
            {type: "aria-haspopup", value: "dialog"},
            {type: "aria-controls", value: 'note-' + note.noteStart},
            {type: "class", value: "timeline-event timeline-note"}
        ]);

        setAttributesTo(button.firstChild, [{type: "class", value: "fas fa-sticky-note"}]);

        initializeModal(dialog, button, closeButton);

        return noteDisplay;
    }

    function formatNote(note) {
        return "<div class='note'>" +
            new showdown.Converter({simplifiedAutoLink: true}).makeHtml('#### ' + formatISODateTime(LocalDateTime.parse(note.noteStart)) + '\n\n' + note.noteContent) +
            "</div>";
    }

    function noteLeftMargin(note) {
        let bulletGap = 8;
        if (note !== notesData[0]) {
            previousTime = LocalDateTime.parse(notesData[notesData.indexOf(note) - 1].noteStart);
            bulletGap = 16;
        }
        return (previousTime.until(LocalDateTime.parse(note.noteStart), ChronoUnit.MINUTES) / THIRTY_MINUTES) * HALF_HOUR_WIDTH - bulletGap;
    }

    function initializeModal(dialog, button, closeButton) {
        const open = function () {
            dialog.setAttribute('aria-hidden', false);
            mainNode.setAttribute('aria-hidden', true);
        };

        const close = function () {
            dialog.setAttribute('aria-hidden', true);
            mainNode.setAttribute('aria-hidden', false);
        };

        button.addEventListener('click', (event) => {
            event.preventDefault();
            open();
        });

        closeButton.addEventListener('click', (event) => {
            close();
        });
    }
}