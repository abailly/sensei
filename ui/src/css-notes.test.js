import {dom} from './dom.js';
import {config} from "./config";

import {describe, expect, test} from '@jest/globals';
import {drawCssNotes} from "./css-notes";

const notes = [{
    "noteStart": "2020-11-16T09:30:00",
    "noteContent": "# Note\r## Première note"
}, {
    "noteStart": "2020-11-16T10:55:00",
    "noteContent": "# Note\r## Seconde note"
}, {
    "noteStart": "2020-11-16T12:20:00",
    "noteContent": "# Note\r## Troisième note"
}];

describe('CSS Notes', () => {

    let mainContainer;
    let container;

    beforeEach(() => {
        config.userProfile = {};
        container = document.createElement("div");
        mainContainer = <div id="main" />;
        container.appendChild(<div/>);
        mainContainer.appendChild(container);
        document.body.appendChild(mainContainer);
    });

    afterEach(() => {
        document.body.removeChild(mainContainer);
    });

    test('expect notes to be drawn', () => {
        config.userProfile.userStartOfDay = '08:30:00';
        config.userProfile.userEndOfDay = '17:00:00';

        drawCssNotes(container, '2020-12-16', notes);

        expect(document.body).toMatchSnapshot();
    });

    test('expect previous drawn notes to be removed', () => {
        config.userProfile.userStartOfDay = '08:30:00';
        config.userProfile.userEndOfDay = '17:00:00';

        drawCssNotes(container, '2020-12-16', notes);
        drawCssNotes(container, '2020-12-16', notes);

        expect(document.body).toMatchSnapshot();
    });
});