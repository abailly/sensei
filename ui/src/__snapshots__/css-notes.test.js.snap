// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CSS Notes expect notes to be drawn 1`] = `
<body>
  <div
    id="main"
  >
    <div>
      <div>
        <div
          class="timeline-title"
          id="title-notes-2020-12-16"
        >
          Notes
        </div>
        <div
          class="timeline-chart"
          id="notes-2020-12-16"
        >
          <ul>
            <li
              style="width: 16px; margin-left: 192px;"
            >
              <div
                aria-controls="note-2020-11-16T09:30:00"
                aria-haspopup="dialog"
                class="timeline-event timeline-note"
                id="note-button-2020-11-16T09:30:00"
                type="button"
              >
                <i
                  class="fas fa-sticky-note"
                />
              </div>
            </li>
            <li
              style="width: 16px; margin-left: 267.33333333333337px;"
            >
              <div
                aria-controls="note-2020-11-16T10:55:00"
                aria-haspopup="dialog"
                class="timeline-event timeline-note"
                id="note-button-2020-11-16T10:55:00"
                type="button"
              >
                <i
                  class="fas fa-sticky-note"
                />
              </div>
            </li>
            <li
              style="width: 16px; margin-left: 267.33333333333337px;"
            >
              <div
                aria-controls="note-2020-11-16T12:20:00"
                aria-haspopup="dialog"
                class="timeline-event timeline-note"
                id="note-button-2020-11-16T12:20:00"
                type="button"
              >
                <i
                  class="fas fa-sticky-note"
                />
              </div>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
  <div
    aria-describedby="note-desc-2020-11-16T12:20:00"
    aria-hidden="true"
    aria-modal="true"
    class="c-dialog"
    id="note-2020-11-16T12:20:00"
    role="[object HTMLDivElement]"
    tabindex="-1"
  >
    <div
      class="c-dialog__box"
    >
      <div
        aria-dismiss="dialog"
        aria-label="Close"
        type="button"
      >
        X
      </div>
      <div
        id="note-desc-2020-11-16T12:20:00"
      >
        <div
          class="note"
        >
          <h4
            id="202011161220"
          >
            2020-11-16 12:20
          </h4>
          

          <h1
            id="note"
          >
            Note
          </h1>
          

          <h2
            id="troisimenote"
          >
            Troisième note
          </h2>
        </div>
      </div>
    </div>
  </div>
  <div
    aria-describedby="note-desc-2020-11-16T10:55:00"
    aria-hidden="true"
    aria-modal="true"
    class="c-dialog"
    id="note-2020-11-16T10:55:00"
    role="[object HTMLDivElement]"
    tabindex="-1"
  >
    <div
      class="c-dialog__box"
    >
      <div
        aria-dismiss="dialog"
        aria-label="Close"
        type="button"
      >
        X
      </div>
      <div
        id="note-desc-2020-11-16T10:55:00"
      >
        <div
          class="note"
        >
          <h4
            id="202011161055"
          >
            2020-11-16 10:55
          </h4>
          

          <h1
            id="note"
          >
            Note
          </h1>
          

          <h2
            id="secondenote"
          >
            Seconde note
          </h2>
        </div>
      </div>
    </div>
  </div>
  <div
    aria-describedby="note-desc-2020-11-16T09:30:00"
    aria-hidden="true"
    aria-modal="true"
    class="c-dialog"
    id="note-2020-11-16T09:30:00"
    role="[object HTMLDivElement]"
    tabindex="-1"
  >
    <div
      class="c-dialog__box"
    >
      <div
        aria-dismiss="dialog"
        aria-label="Close"
        type="button"
      >
        X
      </div>
      <div
        id="note-desc-2020-11-16T09:30:00"
      >
        <div
          class="note"
        >
          <h4
            id="202011160930"
          >
            2020-11-16 09:30
          </h4>
          

          <h1
            id="note"
          >
            Note
          </h1>
          

          <h2
            id="premirenote"
          >
            Première note
          </h2>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`CSS Notes expect previous drawn notes to be removed 1`] = `
<body>
  <div
    id="main"
  >
    <div>
      <div>
        <div
          class="timeline-title"
          id="title-notes-2020-12-16"
        >
          Notes
        </div>
        <div
          class="timeline-chart"
          id="notes-2020-12-16"
        >
          <ul>
            <li
              style="width: 16px; margin-left: 192px;"
            >
              <div
                aria-controls="note-2020-11-16T09:30:00"
                aria-haspopup="dialog"
                class="timeline-event timeline-note"
                id="note-button-2020-11-16T09:30:00"
                type="button"
              >
                <i
                  class="fas fa-sticky-note"
                />
              </div>
            </li>
            <li
              style="width: 16px; margin-left: 267.33333333333337px;"
            >
              <div
                aria-controls="note-2020-11-16T10:55:00"
                aria-haspopup="dialog"
                class="timeline-event timeline-note"
                id="note-button-2020-11-16T10:55:00"
                type="button"
              >
                <i
                  class="fas fa-sticky-note"
                />
              </div>
            </li>
            <li
              style="width: 16px; margin-left: 267.33333333333337px;"
            >
              <div
                aria-controls="note-2020-11-16T12:20:00"
                aria-haspopup="dialog"
                class="timeline-event timeline-note"
                id="note-button-2020-11-16T12:20:00"
                type="button"
              >
                <i
                  class="fas fa-sticky-note"
                />
              </div>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
  <div
    aria-describedby="note-desc-2020-11-16T12:20:00"
    aria-hidden="true"
    aria-modal="true"
    class="c-dialog"
    id="note-2020-11-16T12:20:00"
    role="[object HTMLDivElement]"
    tabindex="-1"
  >
    <div
      class="c-dialog__box"
    >
      <div
        aria-dismiss="dialog"
        aria-label="Close"
        type="button"
      >
        X
      </div>
      <div
        id="note-desc-2020-11-16T12:20:00"
      >
        <div
          class="note"
        >
          <h4
            id="202011161220"
          >
            2020-11-16 12:20
          </h4>
          

          <h1
            id="note"
          >
            Note
          </h1>
          

          <h2
            id="troisimenote"
          >
            Troisième note
          </h2>
        </div>
      </div>
    </div>
  </div>
  <div
    aria-describedby="note-desc-2020-11-16T10:55:00"
    aria-hidden="true"
    aria-modal="true"
    class="c-dialog"
    id="note-2020-11-16T10:55:00"
    role="[object HTMLDivElement]"
    tabindex="-1"
  >
    <div
      class="c-dialog__box"
    >
      <div
        aria-dismiss="dialog"
        aria-label="Close"
        type="button"
      >
        X
      </div>
      <div
        id="note-desc-2020-11-16T10:55:00"
      >
        <div
          class="note"
        >
          <h4
            id="202011161055"
          >
            2020-11-16 10:55
          </h4>
          

          <h1
            id="note"
          >
            Note
          </h1>
          

          <h2
            id="secondenote"
          >
            Seconde note
          </h2>
        </div>
      </div>
    </div>
  </div>
  <div
    aria-describedby="note-desc-2020-11-16T09:30:00"
    aria-hidden="true"
    aria-modal="true"
    class="c-dialog"
    id="note-2020-11-16T09:30:00"
    role="[object HTMLDivElement]"
    tabindex="-1"
  >
    <div
      class="c-dialog__box"
    >
      <div
        aria-dismiss="dialog"
        aria-label="Close"
        type="button"
      >
        X
      </div>
      <div
        id="note-desc-2020-11-16T09:30:00"
      >
        <div
          class="note"
        >
          <h4
            id="202011160930"
          >
            2020-11-16 09:30
          </h4>
          

          <h1
            id="note"
          >
            Note
          </h1>
          

          <h2
            id="premirenote"
          >
            Première note
          </h2>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`CSS Timeline expect notes to be drawn 1`] = `
<body>
  <div
    id="main"
  >
    <div>
      <div>
        <div
          class="timeline-header"
          id="title-notes-2020-12-16"
        >
          Notes
        </div>
        <div
          class="timeline-chart"
          id="notes-2020-12-16"
        >
          <ul>
            <li
              style="width: 16px; margin-left: 192px;"
            >
              <div
                aria-controls="note-2020-11-16T09:30:00"
                aria-haspopup="dialog"
                class="timeline-event timeline-note"
                id="note-button-2020-11-16T09:30:00"
                type="button"
              >
                <i
                  class="fas fa-sticky-note"
                />
              </div>
            </li>
            <li
              style="width: 16px; margin-left: 267.33333333333337px;"
            >
              <div
                aria-controls="note-2020-11-16T10:55:00"
                aria-haspopup="dialog"
                class="timeline-event timeline-note"
                id="note-button-2020-11-16T10:55:00"
                type="button"
              >
                <i
                  class="fas fa-sticky-note"
                />
              </div>
            </li>
            <li
              style="width: 16px; margin-left: 267.33333333333337px;"
            >
              <div
                aria-controls="note-2020-11-16T12:20:00"
                aria-haspopup="dialog"
                class="timeline-event timeline-note"
                id="note-button-2020-11-16T12:20:00"
                type="button"
              >
                <i
                  class="fas fa-sticky-note"
                />
              </div>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
  <div
    aria-describedby="note-desc-2020-11-16T12:20:00"
    aria-hidden="true"
    aria-modal="true"
    class="c-dialog"
    id="note-2020-11-16T12:20:00"
    role="[object HTMLDivElement]"
    tabindex="-1"
  >
    <div
      class="c-dialog__box"
    >
      <div
        aria-dismiss="dialog"
        aria-label="Close"
        type="button"
      >
        X
      </div>
      <div
        id="note-desc-2020-11-16T12:20:00"
      >
        <div
          class="note"
        >
          <h4
            id="202011161220"
          >
            2020-11-16 12:20
          </h4>
          

          <h1
            id="note"
          >
            Note
          </h1>
          

          <h2
            id="troisimenote"
          >
            Troisième note
          </h2>
        </div>
      </div>
    </div>
  </div>
  <div
    aria-describedby="note-desc-2020-11-16T10:55:00"
    aria-hidden="true"
    aria-modal="true"
    class="c-dialog"
    id="note-2020-11-16T10:55:00"
    role="[object HTMLDivElement]"
    tabindex="-1"
  >
    <div
      class="c-dialog__box"
    >
      <div
        aria-dismiss="dialog"
        aria-label="Close"
        type="button"
      >
        X
      </div>
      <div
        id="note-desc-2020-11-16T10:55:00"
      >
        <div
          class="note"
        >
          <h4
            id="202011161055"
          >
            2020-11-16 10:55
          </h4>
          

          <h1
            id="note"
          >
            Note
          </h1>
          

          <h2
            id="secondenote"
          >
            Seconde note
          </h2>
        </div>
      </div>
    </div>
  </div>
  <div
    aria-describedby="note-desc-2020-11-16T09:30:00"
    aria-hidden="true"
    aria-modal="true"
    class="c-dialog"
    id="note-2020-11-16T09:30:00"
    role="[object HTMLDivElement]"
    tabindex="-1"
  >
    <div
      class="c-dialog__box"
    >
      <div
        aria-dismiss="dialog"
        aria-label="Close"
        type="button"
      >
        X
      </div>
      <div
        id="note-desc-2020-11-16T09:30:00"
      >
        <div
          class="note"
        >
          <h4
            id="202011160930"
          >
            2020-11-16 09:30
          </h4>
          

          <h1
            id="note"
          >
            Note
          </h1>
          

          <h2
            id="premirenote"
          >
            Première note
          </h2>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`CSS Timeline expect previous drawn notes to be removed 1`] = `
<body>
  <div
    id="main"
  >
    <div>
      <div>
        <div
          class="timeline-header"
          id="title-notes-2020-12-16"
        >
          Notes
        </div>
        <div
          class="timeline-chart"
          id="notes-2020-12-16"
        >
          <ul>
            <li
              style="width: 16px; margin-left: 192px;"
            >
              <div
                aria-controls="note-2020-11-16T09:30:00"
                aria-haspopup="dialog"
                class="timeline-event timeline-note"
                id="note-button-2020-11-16T09:30:00"
                type="button"
              >
                <i
                  class="fas fa-sticky-note"
                />
              </div>
            </li>
            <li
              style="width: 16px; margin-left: 267.33333333333337px;"
            >
              <div
                aria-controls="note-2020-11-16T10:55:00"
                aria-haspopup="dialog"
                class="timeline-event timeline-note"
                id="note-button-2020-11-16T10:55:00"
                type="button"
              >
                <i
                  class="fas fa-sticky-note"
                />
              </div>
            </li>
            <li
              style="width: 16px; margin-left: 267.33333333333337px;"
            >
              <div
                aria-controls="note-2020-11-16T12:20:00"
                aria-haspopup="dialog"
                class="timeline-event timeline-note"
                id="note-button-2020-11-16T12:20:00"
                type="button"
              >
                <i
                  class="fas fa-sticky-note"
                />
              </div>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
  <div
    aria-describedby="note-desc-2020-11-16T12:20:00"
    aria-hidden="true"
    aria-modal="true"
    class="c-dialog"
    id="note-2020-11-16T12:20:00"
    role="[object HTMLDivElement]"
    tabindex="-1"
  >
    <div
      class="c-dialog__box"
    >
      <div
        aria-dismiss="dialog"
        aria-label="Close"
        type="button"
      >
        X
      </div>
      <div
        id="note-desc-2020-11-16T12:20:00"
      >
        <div
          class="note"
        >
          <h4
            id="202011161220"
          >
            2020-11-16 12:20
          </h4>
          

          <h1
            id="note"
          >
            Note
          </h1>
          

          <h2
            id="troisimenote"
          >
            Troisième note
          </h2>
        </div>
      </div>
    </div>
  </div>
  <div
    aria-describedby="note-desc-2020-11-16T10:55:00"
    aria-hidden="true"
    aria-modal="true"
    class="c-dialog"
    id="note-2020-11-16T10:55:00"
    role="[object HTMLDivElement]"
    tabindex="-1"
  >
    <div
      class="c-dialog__box"
    >
      <div
        aria-dismiss="dialog"
        aria-label="Close"
        type="button"
      >
        X
      </div>
      <div
        id="note-desc-2020-11-16T10:55:00"
      >
        <div
          class="note"
        >
          <h4
            id="202011161055"
          >
            2020-11-16 10:55
          </h4>
          

          <h1
            id="note"
          >
            Note
          </h1>
          

          <h2
            id="secondenote"
          >
            Seconde note
          </h2>
        </div>
      </div>
    </div>
  </div>
  <div
    aria-describedby="note-desc-2020-11-16T09:30:00"
    aria-hidden="true"
    aria-modal="true"
    class="c-dialog"
    id="note-2020-11-16T09:30:00"
    role="[object HTMLDivElement]"
    tabindex="-1"
  >
    <div
      class="c-dialog__box"
    >
      <div
        aria-dismiss="dialog"
        aria-label="Close"
        type="button"
      >
        X
      </div>
      <div
        id="note-desc-2020-11-16T09:30:00"
      >
        <div
          class="note"
        >
          <h4
            id="202011160930"
          >
            2020-11-16 09:30
          </h4>
          

          <h1
            id="note"
          >
            Note
          </h1>
          

          <h2
            id="premirenote"
          >
            Première note
          </h2>
        </div>
      </div>
    </div>
  </div>
</body>
`;
